File: src/delayed-queue/worker/move-due-jobs.js
const { createTimer, toUnixTimestamp } = require('../utils/timer');
const keys = require('../keys');

/**
 * Creates a task that repeatedly checks 'allJobs' SortedSet
 * and moves due jobs to the 'dueJobs' List
 *
 * The fact that a job is due is determined by 'job.due_time <= currentTime' condition
 *
 * @param {Object} options - task options
 * @param {Object} options.redis - An `ioredis` Redis client instance
 * @param {number} options.moveDueJobsInterval - the interval between recurring invocation of the task (in milliseconds)
 * @param {number} options.maxDueJobsToMove - the max limit of due jobs to move within single turn
 * @param {Function} options.onTick - called after task completion on each interval tick
 */
const moveDueJobs = (options) => {
  const {
    redis,
    moveDueJobsInterval,
    maxDueJobsToMove,
    onTick = () => {}
  } = options;

  /**
   * Start the 'move due jobs' activity
   * Runs the task recurrently at given interval
   *
   * @param {AbortSignal} cancelSignal - cancellation signal
   * @returns {Promise} it's neverending Promise, which however fails on error or when cancelled from the outside
   */
  async function start(cancelSignal) {
    for await (const tick of createTimer(moveDueJobsInterval, { signal: cancelSignal })) {
      await moveDueJobsOnce(tick.time, cancelSignal);
      onTick(tick);
    }
  }

  /**
   * Checks for due jobs in 'allJobs' SortedSet and moves them to 'dueJobs' List
   *
   * @param {number} currentTime
   * @param {AbortSignal} cancelSignal - cancellation signal
   * @returns {Promise}
   */
  async function moveDueJobsOnce(currentTime, cancelSignal) {
    let hasMoreDueJobs = true;

    // call moveDueJobs Lua script, but split it in multiple invocations
    // each invocation is allowed to process up to {options.maxJobsToMove}
    // this is to avoid long-running blocking 'moveDueJobs()' Lua script
    // if there're more jobs to move at currentTime,
    // repeat calling moveDueJobs() until all due jobs are moved
    while (hasMoreDueJobs) {
      hasMoreDueJobs = Boolean(await redis.moveDueJobs(
        2,
        keys.allJobs,
        keys.dueJobs,
        toUnixTimestamp(currentTime),
        maxDueJobsToMove
      ));
      cancelSignal.throwIfAborted();
    }
  }

  return {
    start,
    _moveDueJobsOnce: moveDueJobsOnce
  };
};

module.exports = moveDueJobs;

File: src/delayed-queue/scripts/move-due-jobs.lua
-- Migrate jobs from the 'allJobs' SortedSet to 'dueJobs' List
-- Due jobs are jobs satisfying the condition 'job.due_time <= current_time'
--
-- KEYS[1]: key which points to 'allJobs' SortedSet
-- KEYS[2]: key which points to 'dueJobs' List
-- ARGV[1]: current time
-- ARGV[2]: max number of jobs to migrate
local allJobsKey = KEYS[1]
local dueJobsKey = KEYS[2]
local currentTime = tonumber(ARGV[1])
local limit = tonumber(ARGV[2])
local hasMore = 0

-- take all jobs having their "due_time <= current_time"
local dueJobs = redis.call('ZRANGEBYSCORE', allJobsKey, '-inf', currentTime, 'LIMIT', 0, limit + 1)

-- detect if there're more jobs beyond given max limit
-- if so, remove last extra item from the table
if #dueJobs > limit then
  hasMore = 1
  table.remove(dueJobs)
end

-- move due jobs to the 'dueJobs' LIST
for i = 1, #dueJobs do
  local jobId = dueJobs[i]
  redis.call('LPUSH', dueJobsKey, jobId)
  redis.call('ZREM', allJobsKey, jobId)
end

-- return boolean flag indicating if there're more due jobs to move
return hasMore
